adaptive/utils.py

```diff
@@ -67,3 +68,18 @@ def decorator(method):
         return functools.wraps(other)(method)
 
     return decorator
+
+
+class _RequireAttrsABCMeta(abc.ABCMeta):
+    def __call__(self, *args, **kwargs):
+        obj = super().__call__(*args, **kwargs)
+        for name, type_ in obj.__annotations__.items():
+            try:
+                x = getattr(obj, name)
+                if not isinstance(x, type_):
+                    raise TypeError(
+                        f"The attribute '{name}' is of {type_} instead of {type(x)}."
+                    )
+            except AttributeError:
+                raise AttributeError(f"Required attribute {name} not set in __init__.")
```

Comment:
You didn't implement the changes I suggested.
There are still 2 things wrong with this:
+ the `try` block extends over too much code: you actually want it to extend only over the `x = getattr(obj, name)`
+ the `raise` in the `except` block will give spurious backtrace unless you `raise ... from None` like I suggested.

Once again:


```
try:
    x = getattr(obj, name)
except AttributeError:
    raise AttributeError(f"Required attribute {name} not set in __init__.") from None
else:
    if not isinstance(x, type_):
        raise TypeError(
            f"The attribute '{name}' is of {type_} instead of {type(x)}."
        )
```

Reply:
Also the error message for the TypeError was confusing IMO; I suggest an alternative for that above

Reply:
Oh, sorry, I didn't see https://github.com/python-adaptive/adaptive/pull/222#discussion_r335838905 before.

I've changed the code into what you suggested.